File: advantech/examples/Python/AI_PollingStreamingAI/PollingStreamingAI.py
#!/usr/bin/python
# -*- coding:utf-8 -*-


"""
/*******************************************************************************
Copyright (c) 1983-2021 Advantech Co., Ltd.
********************************************************************************
THIS IS AN UNPUBLISHED WORK CONTAINING CONFIDENTIAL AND PROPRIETARY INFORMATION
WHICH IS THE PROPERTY OF ADVANTECH CORP., ANY DISCLOSURE, USE, OR REPRODUCTION,
WITHOUT WRITTEN AUTHORIZATION FROM ADVANTECH CORP., IS STRICTLY PROHIBITED.

================================================================================
REVISION HISTORY
--------------------------------------------------------------------------------
$Log:  $
--------------------------------------------------------------------------------
$NoKeywords:  $
*/
/******************************************************************************
*
* Windows Example:
*    PollingStramingAI.py
*
* Example Category:
*    AI
*
* Description:
*    This example demonstrates how to use Polling Streaming AI function.
*
* Instructions for Running:
*    1. Set the 'deviceDescription' which can get from system device manager for opening the device.
*    2. Set the 'profilePath' to save the profile path of being initialized device.
*    3. Set the 'startChannel' as the first channel for scan analog samples
*    4. Set the 'channelCount' to decide how many sequential channels to scan analog samples.
*    5. Set the 'sectionLength' as the length of data section for Buffered AI.
*    6. Set the 'sectionCount' as the count of data section for Buffered AI.
*
* I/O Connections Overview:
*    Please refer to your hardware reference manual.
*
******************************************************************************/
"""
import sys
sys.path.append('..')
from CommonUtils import kbhit

from Automation.BDaq import *
from Automation.BDaq.WaveformAiCtrl import WaveformAiCtrl
from Automation.BDaq.BDaqApi import AdxEnumToString, BioFailed

# Configure the following parameters before running the demo
deviceDescription = "DemoDevice,BID#0"
profilePath = u"../../profile/DemoDevice.xml"
startChannel = 0
channelCount = 2
sectionLength = 1024
sectionCount = 0

# user buffer size should be equal or greater than raw data buffer length, because data ready count
# is equal or more than smallest section of raw data buffer and up to raw data buffer length.
# users can set 'USER_BUFFER_SIZE' according to demand.
USER_BUFFER_SIZE = channelCount * sectionLength

def AdvPollingStreamingAI():
    ret = ErrorCode.Success

    # Step 1: Create a 'WaveformAiCtrl' for Buffered AI function
    # Select a device by device number pr device description and specify the access mode.
    # In this example we use ModeWrite mode so that we can use fully control the device,
    # including configuring, sampling, etc
    wfAiCtrl = WaveformAiCtrl(deviceDescription)
    for _ in range(1):
        wfAiCtrl.loadProfile = profilePath   # Loads a profile to initialize the device

        # Step 2: Set necessary parameters for Streaming AI operation
        # get the Conversion instance and set the start channel and scan channel number
        wfAiCtrl.conversion.channelStart = startChannel
        wfAiCtrl.conversion.channelCount = channelCount

        # get the record instance and set record count and section length
        wfAiCtrl.record.sectionCount = sectionCount   # The 0 means setting 'streaming' mode
        wfAiCtrl.record.sectionLength = sectionLength

        # Step 3: The operation has been started
        ret = wfAiCtrl.prepare()
        if BioFailed(ret):
            break

        ret = wfAiCtrl.start()
        if BioFailed(ret):
            break

        # Step 4: The device is acquisition data with Polling Style
        print("Polling infinite acquisition is in progress, any key to quit!")
        while not kbhit():
            result = wfAiCtrl.getDataF64(USER_BUFFER_SIZE, -1)
            ret, returnedCount, data, = result[0], result[1], result[2]
            if BioFailed(ret):
                break
            print("Polling Stream AI get data count is %d" % returnedCount)
            print("the first sample for each channel are:")
            for i in range(channelCount):
                print("channel %d: %10.6f" % (i + startChannel, data[i]))
        # Step 6: Stop the operation if it is running
        ret = wfAiCtrl.stop()

    # Step 7: Close device, release any allocated resource
    wfAiCtrl.dispose()

    # If something wrong in this execution, print the error code on screen for tracking
    if BioFailed(ret):
        enumStr = AdxEnumToString("ErrorCode", ret.value, 256)
        print("Some error occurred. And the last error code is %#x. [%s]" % (ret.value, enumStr))
    return 0


if __name__ == '__main__':
    AdvPollingStreamingAI()

File: advantech/examples/C++_Console/AI_PollingStreamingAI/PollingStreamingAI.cpp
/*******************************************************************************
Copyright (c) 1983-2016 Advantech Co., Ltd.
********************************************************************************
THIS IS AN UNPUBLISHED WORK CONTAINING CONFIDENTIAL AND PROPRIETARY INFORMATION
WHICH IS THE PROPERTY OF ADVANTECH CORP., ANY DISCLOSURE, USE, OR REPRODUCTION,
WITHOUT WRITTEN AUTHORIZATION FROM ADVANTECH CORP., IS STRICTLY PROHIBITED. 

================================================================================
REVISION HISTORY
--------------------------------------------------------------------------------
$Log:  $
--------------------------------------------------------------------------------
$NoKeywords:  $
*/
/******************************************************************************
*
* Windows Example:
*    PollingStramingAI.cpp
*
* Example Category:
*    AI
*
* Description:
*    This example demonstrates how to use Polling Streaming AI function.
*
* Instructions for Running:
*    1. Set the 'deviceDescription' which can get from system device manager for opening the device. 
*	  2. Set the 'profilePath' to save the profile path of being initialized device. 
*    3. Set the 'startChannel' as the first channel for scan analog samples  
*    4. Set the 'channelCount' to decide how many sequential channels to scan analog samples.
*    5. Set the 'sectionLength' as the length of data section for Buffered AI.
*	  6. Set the 'sectionCount' as the count of data section for Buffered AI.
*
* I/O Connections Overview:
*    Please refer to your hardware reference manual.
*
******************************************************************************/
#include <stdlib.h>
#include <stdio.h>
#include "../inc/compatibility.h"
#include "../../../inc/bdaqctrl.h"
using namespace Automation::BDaq;
//-----------------------------------------------------------------------------------
// Configure the following parameters before running the demo
//-----------------------------------------------------------------------------------
#define       deviceDescription  L"DemoDevice,BID#0"
const wchar_t* profilePath = L"../../profile/DemoDevice.xml";
int32         startChannel = 0;
const int32   channelCount = 2;
const int32   sectionLength = 1024;
const int32   sectionCount = 0;		
// user buffer size should be equal or greater than raw data buffer length, because data ready count
// is equal or more than smallest section of raw data buffer and up to raw data buffer length.
// users can set 'USER_BUFFER_SIZE' according to demand.
#define  USER_BUFFER_SIZE  channelCount*sectionLength
double   userDataBuffer[USER_BUFFER_SIZE];
int32 returnedCount = 0;

inline void waitAnyKey()
{
	do {SLEEP(1);} while (!kbhit());
}

int main(int argc, char* argv[])
{
	ErrorCode ret = Success;

	// Step 1: Create a 'WaveformAiCtrl' for Buffered AI function.
	WaveformAiCtrl* wfAiCtrl = WaveformAiCtrl::Create();

	do 
	{
		// Step 2: Select a device by device number or device description and specify the access mode.
		// in this example we use ModeWrite mode so that we can fully control the device, including configuring, sampling, etc.
		DeviceInformation devInfo(deviceDescription);
		ret = wfAiCtrl->setSelectedDevice(devInfo);
		CHK_RESULT(ret);
		ret = wfAiCtrl->LoadProfile(profilePath);//Loads a profile to initialize the device. 
		CHK_RESULT(ret);

		// Step 3: Set necessary parameters. 
		Conversion* conversion = wfAiCtrl->getConversion();
		ret = conversion->setChannelStart(startChannel);
		CHK_RESULT(ret);
		ret = conversion->setChannelCount(channelCount);
		CHK_RESULT(ret);
		Record* record = wfAiCtrl->getRecord();
		ret = record->setSectionLength(sectionLength);
		CHK_RESULT(ret);
		ret = record->setSectionCount(sectionCount);//The 0 means setting 'streaming' mode.
		CHK_RESULT(ret);

		// Step 4: The operation has been started.
		ret = wfAiCtrl->Prepare();
		CHK_RESULT(ret);
		ret = wfAiCtrl->Start();
		CHK_RESULT(ret);

		// Step 5: The device is acquiring data with Polling Style.
		printf("Polling infinite acquisition is in progress.\n");
		do 
		{
			ret = wfAiCtrl->GetData(USER_BUFFER_SIZE, userDataBuffer, -1, &returnedCount);//The timeout value is -1, meaning infinite waiting. 
			CHK_RESULT(ret);
			printf("Polling Streaming AI get data count is  %d\n", returnedCount);
			printf("the first sample for each Channel are:\n");
			for (int32 i = 0; i < channelCount; ++i)
			{
				printf("channel %d:%10.6f \n",(i + startChannel), userDataBuffer[i]);   
			}
		} while(!kbhit());

		// step 6: Stop the operation if it is running.
		ret = wfAiCtrl->Stop();
		CHK_RESULT(ret);
	} while (false);

	// Step 7: Close device, release any allocated resource.
	wfAiCtrl->Dispose();

	// If something wrong in this execution, print the error code on screen for tracking.
	if (BioFailed(ret))
	{
      wchar_t enumString[256];
      AdxEnumToString(L"ErrorCode", (int32)ret, 256, enumString);
      printf("Some error occurred. And the last error code is 0x%X. [%ls]\n", ret, enumString);
		waitAnyKey();// wait any key to quit!
	}
	return 0;
}
